turbinia/workers/docker.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Task for running docker-explorer."""
+
+from __future__ import unicode_literals
+
+import json
+import logging
+import os
+import subprocess
+
+from turbinia import TurbiniaException
+from turbinia.evidence import DockerContainer
+from turbinia.processors import mount_local
+from turbinia.workers import TurbiniaTask
+
+log = logging.getLogger('turbinia')
+
+
+class DockerContainersEnumerationTask(TurbiniaTask):
+  """Enumerates Docker containers on Linux"""
+
+  def GetContainers(self, evidence):
+    """Lists the containers from an input Evidence.
+
+    Args:
+      evidence: the input Evidence.
+
+    Returns:
+      a list(dict) containing information about the containers found.
+
+    Raises:
+      TurbiniaException: when the docker-explorer tool failed to run.
+    """
+
+    mount_path = evidence.local_path
+    if type(evidence).__name__ == 'RawDisk':
+      mount_path = mount_local.PreprocessMountDisk(
+          evidence.loopdevice_path, evidence.mount_partition)
+
+    docker_dir = os.path.join(mount_path, 'var', 'lib', 'docker')
+    containers_info = None
+    docker_explorer_command = [
+        'sudo', '/usr/local/bin/de.py', '-r', docker_dir, 'list',
+        'all_containers'
+    ]
+    try:
+      log.info('Running {0:s}'.format(' '.join(docker_explorer_command)))
+      json_string = subprocess.check_output(docker_explorer_command)
+    except Exception as e:
+      mount_local.PostprocessUnmountPath(mount_path)
+      raise TurbiniaException(
+          'Failed to run {0:s} {1!s}'.format(
+              ' '.join(docker_explorer_command), e))
+
+    try:
+      containers_info = json.loads(json_string)
+    except ValueError as e:
+      mount_local.PostprocessUnmountPath(mount_path)
```

Comment:
Here are a couple more examples of post-processors running in the `run()` method instead of being called implicitly\automatically by the run wrapper.  Same as above, but I'm not sure I understand why that's being done here.

Reply:
Since I have to manually mount the underlying partition of parent RawDisk image (which preprocessor doesn't do), I have to clean up afterwards.

Reply:
This same thing is going to happen with the GoogleCloudDisk AFAICT.  I think we should actually move this into the RawDisk preprocessor (and the GoogleCloudDisk preprocessor), and have a property of the disk based evidence types that is something like `mounted` that the preprocessors can set when they successfully mount the partition(s).  Then Tasks that care whether the disks are mounted are not can check for this and bail out if there is a problem, but for those Tasks that don't care (like Plaso), they can just ignore this.   As mentioned above, at a later time we can refactor this into two different evidence types (RawDisk and RawDiskPartition or maybe RawDiskVolume).

Another step for later will be disentangling the preprocessors for the other evidence types (e.g. GoogleCloudDisk) so that they either set and use the `parent_evidence` or use inheritance so that there is not code duplication for each one.

Reply:
moved mounting/unmounting in rawdisk